torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -225,8 +225,16 @@ void Reducer::mark_variable_ready_dense(VariableIndex index) {
     // `detach_` from `zero_grad`, which is incompatible with views.
     TORCH_INTERNAL_ASSERT(!grad.is_alias_of(bucket_view));
     TORCH_INTERNAL_ASSERT(grad.device() == bucket_view.device());
-    TORCH_INTERNAL_ASSERT(grad.numel() == bucket_view.numel());
-    bucket_view.copy_(grad.view({-1}), /* non_blocking */ true);
+    TORCH_INTERNAL_ASSERT(grad.sizes() == bucket_view.sizes());
+    // Checks below are too harsh.  AccumulateGrad doesn't HAVE to obey
+    // the grad layout contract.  The penalty for disobedience is reduced
+    // performance, not numerical death.
+    // What should I do here?  Warn?  Do nothing?
```

Comment:
The copy_ will handle it properly right? So I guess nothing?
Note that the view, if you add it back, might not handle it though. So you might have to use reshape here.

Reply:
The view comment is moot because both have the same size.
So I would guess nothing here?

Reply:
Welll maybe i should warn that grad strides are breaking the contract and a perf regression may occur.  If someone complains to us about poor optimizer or DDP performance, these warnings would be a useful data point to help us debug (certainly better than no warning at all), even if they don't tell the user exactly what they should do to fix it.

Reply:
Yes sounds good, a `TORCH_WARN_ONCE` sounds good here.

Reply:
I decided to split the two warnings.
Reducer.cpp warns if [grad strides != bucket view strides](https://github.com/pytorch/pytorch/pull/34904/files#diff-c269190a925a4b0df49eda8a8f6c5bd3R233-R239).
accumulate_grad.h warns if [incoming grad and variable do not obey the layout contract](https://github.com/pytorch/pytorch/pull/34904/files#diff-ef1a56d24f66b280dcdb401502d6a796R150-R155).
(just realized there is a typo in sparseness checking, should be `if (!(new_grad.is_sparse() || variable_grad.is_sparse())) {`, fixed locally)
What do you think?